### DISCUSSION TITLE
XmppComponent

### DISCUSSION THREAD

**Discussion by jason-tomlinson** [vitalyster/SharpXMPP#209] [Q&A]
Hello,

For XmppComponent, the spec requires that the To field be the component name.
( to="componentname.domain" )

Could XmppTcpConnection be modified to use this format when connecting with the 'jabber:component:accept' namespace?

**Comment by vitalyster** [comment]
You should pass component JID to XmppComponent constructor, "componentname.domain" is the component JID and component will send it correctly

**Comment by jason-tomlinson** [comment]
Thanks for the advice.
But, if the JID is created as 
  `var j = new JID("component name.domain");` 
then the domain is treated as "component name.domain" and is not known.

If the JID is created as 
 `var j = new JID("component name@domain");` 
then the domain is reached, but the format is incorrect and I receive
```
<handshake xmlns="jabber:component:accept">552aa9defce51a940be725e6cf39c5d7d2e517d1</handshake>

<stream:error xmlns:stream="http://etherx.jabber.org/streams">
  <host-unknown xmlns="urn:ietf:params:xml:ns:xmpp-streams" />
</stream:error>
```

If I modify XmppTcpConnection to use the first style in the 'to' field, then it is OK and the server returns the completed handshake. (For info, this is with Tigase)

Or is there another JID class for components only that I missed?

**Reply by vitalyster** [reply]
> the domain is treated as "component name.domain" and is not known.

"componentname.domain" is the JID and the domain of the component. Not clear what is wrong. `new JID('componentname.domain')` will be sent as `to="componentname.domain"`. Or you see something different?

**Comment by jason-tomlinson** [comment]
Well, in that jid the domain contains componentname, too ` jid.Domain='componentname.domain'`, which cannot be resolved by ResolveHostAddresses 
Same as ping XXXXX.google.com wont work.

I also had to make another couple of changes - after the handshake is completed, that is all the authentication that is done (no sasl features are advertised, nor currently supported, https://xmpp.org/extensions/xep-0114.html#proto ).
So in XmppTcpConnection.StartAuthentication no SASLHandler should be created for components (the authentication is already done).
And in InitTlsIfSupported features are likely to be null because of this, so it must return false (or handled differently)

**Reply by vitalyster** [reply]
Your component name should be resolved to domain address and should have SRV record as well as main domain. If ResolveHostAddresses does not find any SRV records it fallsback to A records with Dns.GetHostAddressesAsync and this domain must actually resolves to some address (127.0.0.1 should be ok). If your setup does not resolve component domain to address then it is incorrect.
